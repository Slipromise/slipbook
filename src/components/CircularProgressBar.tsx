import React, { HTMLAttributes, useMemo } from "react";
import styles from "@/styles/components/CircularProgressBar.module.scss";
import { AnimationConfig, useSpring, animated } from "@react-spring/web";

type Props = {
  value: number;
  title?: string;
  color?: string;
  duration?: number;
};

function CircularProgressBar({
  value = 0,
  title,
  color,
  duration = 400,
}: Props) {
  const containerStyle = useMemo<
    HTMLAttributes<HTMLDivElement>["style"]
  >(() => {
    const percent = Math.floor(value);

    return {
      "--percent": `${percent}%`,
      "--color": color,
      "--duration": `${duration}ms`,
    } as HTMLAttributes<HTMLDivElement>["style"];
  }, [color, duration, value]);

  const animatedValues = useSpring<{ percent: number }>({
    percent: value,
    config: {
      duration,
    } as AnimationConfig,
  });

  return (
    <div className={styles.container} style={containerStyle}>
      <animated.h1>
        {animatedValues.percent.to((n) => Math.floor(n))}
      </animated.h1>
      {title && <h2>{title}</h2>}
    </div>
  );
}

export default CircularProgressBar;
