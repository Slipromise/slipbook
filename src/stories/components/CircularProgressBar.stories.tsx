import CircularProgressBar from "@/components/CircularProgressBar";
import type { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof CircularProgressBar> = {
  component: CircularProgressBar,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof CircularProgressBar>;

export const Normal: Story = {
  args: {
    value: 40,
  },
};

export const WithTitles: Story = {
  args: {
    value: 50,
    title: "React",
  },
};

export const Colorize: Story = {
  args: {
    value: 60,
    title: "React",
    color: "red",
  },
};
