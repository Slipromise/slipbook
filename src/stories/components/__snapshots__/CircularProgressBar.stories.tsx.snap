// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/CircularProgressBar Colorize smoke-test 1`] = `
<div class="_container_4lpsq_6"
     style="--percent: 60%; --color: red; --duration: 400ms;"
>
  <h1>
    60
  </h1>
  <h2>
    React
  </h2>
</div>
`;

exports[`stories/components/CircularProgressBar Normal smoke-test 1`] = `
<div class="_container_4lpsq_6"
     style="--percent: 40%; --duration: 400ms;"
>
  <h1>
    40
  </h1>
</div>
`;

exports[`stories/components/CircularProgressBar WithTitles smoke-test 1`] = `
<div class="_container_4lpsq_6"
     style="--percent: 50%; --duration: 400ms;"
>
  <h1>
    50
  </h1>
  <h2>
    React
  </h2>
</div>
`;
